File: dotnet/src/OperationVanguard.Core/Simulation/Spawn.cs
namespace OperationVanguard.Core;

/// <summary>A transient hazard the spawn system should route players away from.</summary>
public sealed class DangerZone
{
    public Vec3 Position { get; set; } = new();

    public double Radius { get; set; }

    public double TimeRemaining { get; set; }
}

public sealed class RecentSpawnDeath
{
    public Vec3 Position { get; set; } = new();

    public double Time { get; set; }
}

/// <summary>Bookkeeping retained between spawn selections.</summary>
public sealed class SpawnContext
{
    public Dictionary<int, double> RecentUse { get; } = [];

    public List<RecentSpawnDeath> RecentDeaths { get; } = [];

    public List<DangerZone> DangerZones { get; } = [];

    public Dictionary<string, double> GroupWeights { get; } = new(StringComparer.Ordinal);
}

public sealed class SpawnChoice
{
    public Vec3 Position { get; set; } = new();

    public double Yaw { get; set; }

    public int Index { get; set; }

    public double Score { get; set; }
}

public sealed class SpawnSelectionOptions
{
    public bool InitialOnly { get; set; }
}

/// <summary>Influence-map spawn scoring and selection.</summary>
public static class SpawnSystem
{
    private static readonly QueryFilter SpawnFilter = new(CollisionLayer.Movement);
    private static readonly QueryFilter SightFilter = new(CollisionLayer.Sight);
    private static readonly Vec3 Eye = new();
    private static readonly Vec3 Forward = new();
    private static readonly Vec3 Candidate = new();

    public static SpawnContext CreateSpawnContext() => new();

    public static void NoteDeath(SpawnContext context, Vec3 position, double time)
    {
        context.RecentDeaths.Add(new RecentSpawnDeath
        {
            Position = new Vec3(position.X, position.Y, position.Z),
            Time = time,
        });

        while (context.RecentDeaths.Count > 48)
        {
            context.RecentDeaths.RemoveAt(0);
        }
    }

    public static void AddDangerZone(
        SpawnContext context,
        Vec3 position,
        double radius,
        double duration)
    {
        context.DangerZones.Add(new DangerZone
        {
            Position = new Vec3(position.X, position.Y, position.Z),
            Radius = radius,
            TimeRemaining = duration,
        });
    }

    public static void TickSpawnContext(SpawnContext context, double deltaTime, double time)
    {
        for (var index = context.DangerZones.Count - 1; index >= 0; index--)
        {
            var zone = context.DangerZones[index];
            zone.TimeRemaining -= deltaTime;
            if (zone.TimeRemaining <= 0d)
            {
                context.DangerZones.RemoveAt(index);
            }
        }

        var cutoff = time - GameConstants.Spawn.RecentUseWindow * 2d;
        while (context.RecentDeaths.Count > 0 && context.RecentDeaths[0].Time < cutoff)
        {
            context.RecentDeaths.RemoveAt(0);
        }
    }

    /// <summary>Score one candidate. Higher is better; negative infinity is unusable.</summary>
    public static double ScoreSpawn(
        WorldState world,
        ICollisionWorld collision,
        SpawnContext context,
        SpawnPoint spawn,
        int spawnIndex,
        PlayerState player,
        double time)
    {
        if (spawn.Team != Team.None && spawn.Team != player.Team)
        {
            return double.NegativeInfinity;
        }

        MathEx.Set(Candidate, spawn.Position.X, spawn.Position.Y, spawn.Position.Z);
        var groundY = collision.GroundHeightAt(Candidate.X, Candidate.Z, Candidate.Y + 3d, 12d);
        if (!double.IsFinite(groundY))
        {
            return double.NegativeInfinity;
        }

        Candidate.Y = groundY + 0.05d;
        if (!collision.IsCapsuleFree(
                Candidate,
                GameConstants.StanceHeight.Stand,
                GameConstants.PlayerRadius,
                SpawnFilter))
        {
            return double.NegativeInfinity;
        }

        var score = 1000d;
        score += context.GroupWeights.GetValueOrDefault(spawn.Group) * 400d;
        score += (spawn.Priority ?? 0d) * 40d;

        var eyeY = Candidate.Y + 1.6d;
        foreach (var other in world.Players.Values)
        {
            if (!other.Alive || other.Id == player.Id)
            {
                continue;
            }

            var distance = MathEx.Distance(Candidate, other.Position);
            if (SimulationTypes.IsEnemyTeam(player.Team, other.Team))
            {
                if (distance < GameConstants.Spawn.EnemyHardBanRadius)
                {
                    return double.NegativeInfinity;
                }

                if (distance < GameConstants.Spawn.EnemyDangerRadius)
                {
                    var amount = 1d - MathEx.Clamp01(
                        (distance - GameConstants.Spawn.EnemyHardBanRadius) /
                        (GameConstants.Spawn.EnemyDangerRadius - GameConstants.Spawn.EnemyHardBanRadius));
                    score -= 700d * amount * amount;
                }

                if (distance < 70d)
                {
                    MathEx.Set(Eye, other.Position.X, other.Position.Y + 1.6d, other.Position.Z);
                    MathEx.AnglesToForward(Forward, other.Yaw, other.Pitch);
                    MathEx.Set(Candidate, Candidate.X, eyeY, Candidate.Z);
                    if (MathEx.InCone(
                            Eye,
                            Forward,
                            Candidate,
                            GameConstants.Spawn.EnemyViewConeHalfAngle,
                            70d) &&
                        collision.IsVisible(Eye, Candidate, SightFilter))
                    {
                        score -= GameConstants.Spawn.EnemyViewConePenalty *
                            (1d - MathEx.Clamp01(distance / 70d) * 0.6d);
                    }

                    Candidate.Y = groundY + 0.05d;
                }
            }
            else
            {
                if (distance < GameConstants.Spawn.FriendlyAttractRadius)
                {
                    score += 90d *
                        (1d - MathEx.Clamp01(distance / GameConstants.Spawn.FriendlyAttractRadius));
                }

                if (distance < 1.6d)
                {
                    score -= 300d;
                }
            }
        }

        if (context.RecentUse.TryGetValue(spawnIndex, out var lastUsed))
        {
            var age = time - lastUsed;
            if (age < GameConstants.Spawn.RecentUseWindow)
            {
                score -= GameConstants.Spawn.RecentUsePenalty *
                    (1d - age / GameConstants.Spawn.RecentUseWindow);
            }
        }

        foreach (var death in context.RecentDeaths)
        {
            var age = time - death.Time;
            if (age > GameConstants.Spawn.RecentUseWindow)
            {
                continue;
            }

            var distance = MathEx.DistanceXz(Candidate, death.Position);
            if (distance < GameConstants.Spawn.RecentDeathRadius)
            {
                var proximity = 1d - distance / GameConstants.Spawn.RecentDeathRadius;
                var recency = 1d - age / GameConstants.Spawn.RecentUseWindow;
                score -= GameConstants.Spawn.RecentDeathPenalty * proximity * recency;
            }
        }

        foreach (var zone in context.DangerZones)
        {
            var distance = MathEx.Distance(Candidate, zone.Position);
            if (distance < zone.Radius)
            {
                score -= GameConstants.Spawn.DangerZonePenalty * (1d - distance / zone.Radius);
            }
        }

        return score;
    }

    public static SpawnChoice? SelectSpawn(
        WorldState world,
        MapDef map,
        ICollisionWorld collision,
        SpawnContext context,
        PlayerState player,
        Rng rng,
        SpawnSelectionOptions? options = null)
    {
        options ??= new SpawnSelectionOptions();
        var best = double.NegativeInfinity;
        var candidates = new List<SpawnChoice>();

        for (var index = 0; index < map.Spawns.Count; index++)
        {
            var spawn = map.Spawns[index];
            if (spawn.InitialOnly == true && !options.InitialOnly)
            {
                continue;
            }

            if (options.InitialOnly && spawn.InitialOnly != true && spawn.Team != player.Team)
            {
                continue;
            }

            var score = ScoreSpawn(world, collision, context, spawn, index, player, world.Time);
            if (double.IsNegativeInfinity(score))
            {
                continue;
            }

            if (score > best)
            {
                best = score;
            }

            candidates.Add(new SpawnChoice
            {
                Position = new Vec3(spawn.Position.X, spawn.Position.Y, spawn.Position.Z),
                Yaw = spawn.Yaw,
                Index = index,
                Score = score,
            });
        }

        if (candidates.Count == 0)
        {
            return FallbackSpawn(map, collision, player);
        }

        var band = Math.Max(120d, Math.Abs(best) * 0.08d);
        var shortlist = candidates.Where(candidate => candidate.Score >= best - band).ToArray();
        var chosen = shortlist.Length > 0
            ? shortlist[rng.Int(0, shortlist.Length - 1)]
            : candidates[0];

        context.RecentUse[chosen.Index] = world.Time;
        var groundY = collision.GroundHeightAt(
            chosen.Position.X,
            chosen.Position.Z,
            chosen.Position.Y + 3d,
            12d);
        if (double.IsFinite(groundY))
        {
            chosen.Position.Y = groundY + 0.05d;
        }

        return chosen;
    }

    public static void SetGroupWeights(
        SpawnContext context,
        IReadOnlyDictionary<string, double> weights)
    {
        context.GroupWeights.Clear();
        foreach (var (group, weight) in weights)
        {
            context.GroupWeights[group] = weight;
        }
    }

    public static void ResetSpawnContext(SpawnContext context)
    {
        context.RecentUse.Clear();
        context.RecentDeaths.Clear();
        context.DangerZones.Clear();
        context.GroupWeights.Clear();
    }

    public static double RespawnDelayFor(PlayerState player, double baseDelay, double maximum)
    {
        _ = player;
        return Math.Min(maximum, baseDelay);
    }

    public static double NearestEnemyDistance(WorldState world, Vec3 from, Team team)
    {
        var best = double.PositiveInfinity;
        foreach (var other in world.Players.Values)
        {
            if (!other.Alive || !SimulationTypes.IsEnemyTeam(team, other.Team))
            {
                continue;
            }

            var distance = MathEx.Distance(from, other.Position);
            if (distance < best)
            {
                best = distance;
            }
        }

        return best;
    }

    public static IReadOnlyList<int> PlayersLookingAt(
        WorldState world,
        ICollisionWorld collision,
        Vec3 point,
        Team team)
    {
        var output = new List<int>();
        foreach (var other in world.Players.Values)
        {
            if (!other.Alive || !SimulationTypes.IsEnemyTeam(team, other.Team))
            {
                continue;
            }

            MathEx.Set(Eye, other.Position.X, other.Position.Y + 1.6d, other.Position.Z);
            MathEx.AnglesToForward(Forward, other.Yaw, other.Pitch);
            if (!MathEx.InCone(
                    Eye,
                    Forward,
                    point,
                    GameConstants.Spawn.EnemyViewConeHalfAngle,
                    70d))
            {
                continue;
            }

            if (!collision.IsVisible(Eye, point, SightFilter))
            {
                continue;
            }

            output.Add(other.Id);
        }

        return output;
    }

    private static SpawnChoice? FallbackSpawn(
        MapDef map,
        ICollisionWorld collision,
        PlayerState player)
    {
        SpawnChoice? best = null;
        for (var index = 0; index < map.Spawns.Count; index++)
        {
            var spawn = map.Spawns[index];
            if (spawn.Team != Team.None && spawn.Team != player.Team)
            {
                continue;
            }

            var groundY = collision.GroundHeightAt(
                spawn.Position.X,
                spawn.Position.Z,
                spawn.Position.Y + 3d,
                12d);
            var y = double.IsFinite(groundY) ? groundY + 0.05d : spawn.Position.Y;
            var choice = new SpawnChoice
            {
                Position = new Vec3(spawn.Position.X, y, spawn.Position.Z),
                Yaw = spawn.Yaw,
                Index = index,
                Score = 0d,
            };
            best ??= choice;
        }

        if (best is not null)
        {
            return best;
        }

        if (map.Spawns.Count == 0)
        {
            return null;
        }

        var any = map.Spawns[0];
        return new SpawnChoice
        {
            Position = new Vec3(any.Position.X, any.Position.Y, any.Position.Z),
            Yaw = any.Yaw,
            Index = 0,
            Score = 0d,
        };
    }
}

File: web/src/shared/sim/spawn.ts
/**
 * Spawn selection.
 *
 * COD does not pick a random spawn point — it scores every candidate against an
 * influence map and takes the best one. That is what makes spawns feel fair even
 * though they are, mechanically, teleporting you into an active firefight.
 *
 * The scoring is intentionally readable: every term is a named penalty or bonus
 * with a constant in SPAWN, so tuning spawn feel is a matter of changing numbers
 * rather than rewriting logic. The single most important term by far is the
 * enemy view-cone penalty — being spawned *behind* an enemy is survivable and
 * even exciting; being spawned in their crosshair is not.
 */

import { PLAYER_RADIUS, SPAWN, STANCE_HEIGHT } from '../constants.js';
import {
  anglesToForward,
  clamp01,
  inCone,
  v3distance,
  v3distanceXZ,
  v3set,
  vec3,
  type Vec3,
} from '../math.js';
import type { Rng } from '../rng.js';
import { Team, isEnemyTeam, type PlayerId, type PlayerState, type WorldState } from '../types.js';
import { CollisionLayer, type CollisionWorld, type QueryFilter } from '../collision/collision-types.js';
import type { MapDef, SpawnPoint } from '../map/map-types.js';

const SPAWN_FILTER: QueryFilter = { layers: CollisionLayer.Movement };

const _eye = vec3();
const _forward = vec3();
const _candidate = vec3();

/** A transient hazard the spawn system should route players away from. */
export interface DangerZone {
  position: Vec3;
  radius: number;
  /** Seconds remaining. Expired zones are ignored and pruned by the caller. */
  timeRemaining: number;
}

/** Bookkeeping the spawn system keeps between selections. */
export interface SpawnContext {
  /** Spawn index -> world time it was last used. */
  recentUse: Map<number, number>;
  /** Recent deaths, oldest first. */
  recentDeaths: Array<{ position: Vec3; time: number }>;
  dangerZones: DangerZone[];
  /**
   * Which spawn groups the mode currently prefers, highest weight first.
   * Domination sets this when a flag changes hands, which is what makes spawns
   * flip sides.
   */
  groupWeights: Map<string, number>;
}

export function createSpawnContext(): SpawnContext {
  return {
    recentUse: new Map(),
    recentDeaths: [],
    dangerZones: [],
    groupWeights: new Map(),
  };
}

export function noteDeath(ctx: SpawnContext, position: Vec3, time: number): void {
  ctx.recentDeaths.push({ position: vec3(position.x, position.y, position.z), time });
  // Bound the list; anything older than the death penalty window is irrelevant.
  while (ctx.recentDeaths.length > 48) ctx.recentDeaths.shift();
}

export function addDangerZone(
  ctx: SpawnContext,
  position: Vec3,
  radius: number,
  duration: number,
): void {
  ctx.dangerZones.push({
    position: vec3(position.x, position.y, position.z),
    radius,
    timeRemaining: duration,
  });
}

export function tickSpawnContext(ctx: SpawnContext, dt: number, time: number): void {
  for (let i = ctx.dangerZones.length - 1; i >= 0; i--) {
    const z = ctx.dangerZones[i]!;
    z.timeRemaining -= dt;
    if (z.timeRemaining <= 0) ctx.dangerZones.splice(i, 1);
  }
  // Drop deaths older than the penalty window so the list can't grow unbounded.
  const cutoff = time - SPAWN.recentUseWindow * 2;
  while (ctx.recentDeaths.length > 0 && ctx.recentDeaths[0]!.time < cutoff) {
    ctx.recentDeaths.shift();
  }
}

// ---------------------------------------------------------------------------
// Scoring
// ---------------------------------------------------------------------------

/**
 * Score one candidate. Higher is better; -Infinity means unusable.
 *
 * Exported so the test suite can assert the properties that matter — that a
 * spawn in an enemy's crosshair always loses to one behind them, and that a
 * spawn next to a live enemy is never chosen when any alternative exists.
 */
export function scoreSpawn(
  world: WorldState,
  collision: CollisionWorld,
  ctx: SpawnContext,
  spawn: SpawnPoint,
  spawnIndex: number,
  player: PlayerState,
  time: number,
): number {
  // Team gate. Team.None spawns are usable by anyone (free-for-all and neutral).
  if (spawn.team !== Team.None && spawn.team !== player.team) return -Infinity;

  // Must physically fit.
  v3set(_candidate, spawn.position.x, spawn.position.y, spawn.position.z);
  const groundY = collision.groundHeightAt(_candidate.x, _candidate.z, _candidate.y + 3, 12);
  if (!Number.isFinite(groundY)) return -Infinity;
  _candidate.y = groundY + 0.05;
  if (!collision.isCapsuleFree(_candidate, STANCE_HEIGHT.stand, PLAYER_RADIUS, SPAWN_FILTER)) {
    return -Infinity;
  }

  let score = 1000;

  // Group preference set by the game mode (objective ownership).
  score += (ctx.groupWeights.get(spawn.group) ?? 0) * 400;
  score += (spawn.priority ?? 0) * 40;

  const eyeY = _candidate.y + 1.6;

  for (const other of world.players.values()) {
    if (!other.alive || other.id === player.id) continue;

    const dist = v3distance(_candidate, other.position);

    if (isEnemyTeam(player.team, other.team)) {
      // Hard ban: far too close to be survivable.
      if (dist < SPAWN.enemyHardBanRadius) return -Infinity;

      if (dist < SPAWN.enemyDangerRadius) {
        // Quadratic ramp so the penalty bites sharply as you approach the ban radius.
        const t = 1 - clamp01((dist - SPAWN.enemyHardBanRadius) /
          (SPAWN.enemyDangerRadius - SPAWN.enemyHardBanRadius));
        score -= 700 * t * t;
      }

      // The decisive term: is this spawn inside the enemy's field of view, with
      // line of sight? Being seen at the instant you appear is the worst
      // possible outcome, so it outweighs everything except the hard ban.
      if (dist < 70) {
        v3set(_eye, other.position.x, other.position.y + 1.6, other.position.z);
        anglesToForward(_forward, other.yaw, other.pitch);
        v3set(_candidate, _candidate.x, eyeY, _candidate.z);
        if (inCone(_eye, _forward, _candidate, SPAWN.enemyViewConeHalfAngle, 70)) {
          const sightFilter: QueryFilter = { layers: CollisionLayer.Sight };
          if (collision.isVisible(_eye, _candidate, sightFilter)) {
            // Closer enemies looking at you are worse than distant ones.
            score -= SPAWN.enemyViewConePenalty * (1 - clamp01(dist / 70) * 0.6);
          }
        }
        _candidate.y = groundY + 0.05;
      }
    } else {
      // Friendlies are a mild attractor: spawning with your team is safer and
      // keeps the squad together, but it must never outweigh enemy proximity.
      if (dist < SPAWN.friendlyAttractRadius) {
        score += 90 * (1 - clamp01(dist / SPAWN.friendlyAttractRadius));
      }
      // ...except right on top of them, which looks broken.
      if (dist < 1.6) score -= 300;
    }
  }

  // Don't reuse the same corner repeatedly.
  const lastUsed = ctx.recentUse.get(spawnIndex);
  if (lastUsed !== undefined) {
    const age = time - lastUsed;
    if (age < SPAWN.recentUseWindow) {
      score -= SPAWN.recentUsePenalty * (1 - age / SPAWN.recentUseWindow);
    }
  }

  // Avoid the spot the player just died in, and anywhere else people are dying.
  for (const death of ctx.recentDeaths) {
    const age = time - death.time;
    if (age > SPAWN.recentUseWindow) continue;
    const d = v3distanceXZ(_candidate, death.position);
    if (d < SPAWN.recentDeathRadius) {
      const proximity = 1 - d / SPAWN.recentDeathRadius;
      const recency = 1 - age / SPAWN.recentUseWindow;
      score -= SPAWN.recentDeathPenalty * proximity * recency;
    }
  }

  // Grenades, killstreaks and fire.
  for (const zone of ctx.dangerZones) {
    const d = v3distance(_candidate, zone.position);
    if (d < zone.radius) {
      score -= SPAWN.dangerZonePenalty * (1 - d / zone.radius);
    }
  }

  return score;
}

// ---------------------------------------------------------------------------
// Selection
// ---------------------------------------------------------------------------

export interface SpawnChoice {
  position: Vec3;
  yaw: number;
  index: number;
  score: number;
}

/**
 * Pick the best spawn for a player.
 *
 * Ties are broken by a seeded random draw among the top candidates rather than
 * always taking the single best, so ten players spawning in the same second
 * don't all stack on the same point. The randomness is drawn from the world RNG
 * so client and server agree.
 */
export function selectSpawn(
  world: WorldState,
  map: MapDef,
  collision: CollisionWorld,
  ctx: SpawnContext,
  player: PlayerState,
  rng: Rng,
  opts: { initialOnly?: boolean } = {},
): SpawnChoice | null {
  let best = -Infinity;
  const candidates: SpawnChoice[] = [];

  for (let i = 0; i < map.spawns.length; i++) {
    const spawn = map.spawns[i]!;
    if (spawn.initialOnly && !opts.initialOnly) continue;
    if (opts.initialOnly && !spawn.initialOnly && spawn.team !== player.team) continue;

    const score = scoreSpawn(world, collision, ctx, spawn, i, player, world.time);
    if (score === -Infinity) continue;

    if (score > best) best = score;
    candidates.push({
      position: vec3(spawn.position.x, spawn.position.y, spawn.position.z),
      yaw: spawn.yaw,
      index: i,
      score,
    });
  }

  if (candidates.length === 0) {
    // Every spawn is banned. Rather than refusing to spawn the player — which
    // would stall the match — fall back to the least-bad point ignoring enemies.
    return fallbackSpawn(map, collision, player);
  }

  // Consider anything within a band of the best. The band is proportional so it
  // stays meaningful whether scores are in the hundreds or the thousands.
  const band = Math.max(120, Math.abs(best) * 0.08);
  const shortlist = candidates.filter((c) => c.score >= best - band);
  const chosen = shortlist.length > 0 ? shortlist[rng.int(0, shortlist.length - 1)]! : candidates[0]!;

  ctx.recentUse.set(chosen.index, world.time);

  // Drop the player onto the floor rather than trusting the authored Y.
  const groundY = collision.groundHeightAt(chosen.position.x, chosen.position.z, chosen.position.y + 3, 12);
  if (Number.isFinite(groundY)) chosen.position.y = groundY + 0.05;

  return chosen;
}

/**
 * Last resort when every scored spawn was rejected: take the team's spawn point
 * that is simply furthest from any living enemy. Never returns null if the map
 * has any spawn at all, because failing to spawn is worse than spawning badly.
 */
function fallbackSpawn(map: MapDef, collision: CollisionWorld, player: PlayerState): SpawnChoice | null {
  let best: SpawnChoice | null = null;
  for (let i = 0; i < map.spawns.length; i++) {
    const spawn = map.spawns[i]!;
    if (spawn.team !== Team.None && spawn.team !== player.team) continue;
    const groundY = collision.groundHeightAt(spawn.position.x, spawn.position.z, spawn.position.y + 3, 12);
    const y = Number.isFinite(groundY) ? groundY + 0.05 : spawn.position.y;
    const choice: SpawnChoice = {
      position: vec3(spawn.position.x, y, spawn.position.z),
      yaw: spawn.yaw,
      index: i,
      score: 0,
    };
    if (!best) best = choice;
  }
  if (best) return best;

  // Truly nothing for this team — use any spawn at all.
  const any = map.spawns[0];
  if (!any) return null;
  return { position: vec3(any.position.x, any.position.y, any.position.z), yaw: any.yaw, index: 0, score: 0 };
}

// ---------------------------------------------------------------------------
// Mode integration
// ---------------------------------------------------------------------------

/**
 * Bias spawns toward groups near objectives a team controls.
 *
 * Called by mode logic whenever ownership changes. Group names are free-form
 * strings chosen by the map author; the convention is `<team>_<area>`, and any
 * group whose name contains the area of a controlled objective gets weighted up.
 */
export function setGroupWeights(ctx: SpawnContext, weights: Record<string, number>): void {
  ctx.groupWeights.clear();
  for (const [group, w] of Object.entries(weights)) {
    ctx.groupWeights.set(group, w);
  }
}

/** Reset between rounds so a new round doesn't inherit the last one's history. */
export function resetSpawnContext(ctx: SpawnContext): void {
  ctx.recentUse.clear();
  ctx.recentDeaths.length = 0;
  ctx.dangerZones.length = 0;
  ctx.groupWeights.clear();
}

/** Respawn delay for a player. Flat by design — see MATCH.respawnDelay. */
export function respawnDelayFor(player: PlayerState, base: number, max: number): number {
  void player;
  return Math.min(max, base);
}

/** Distance from a player to the nearest living enemy — used by tests and AI. */
export function nearestEnemyDistance(world: WorldState, from: Vec3, team: Team): number {
  let best = Infinity;
  for (const other of world.players.values()) {
    if (!other.alive) continue;
    if (!isEnemyTeam(team, other.team)) continue;
    const d = v3distance(from, other.position);
    if (d < best) best = d;
  }
  return best;
}

/** Player ids currently looking at a point — used for spawn debugging overlays. */
export function playersLookingAt(
  world: WorldState,
  collision: CollisionWorld,
  point: Vec3,
  team: Team,
): PlayerId[] {
  const out: PlayerId[] = [];
  const sightFilter: QueryFilter = { layers: CollisionLayer.Sight };
  for (const other of world.players.values()) {
    if (!other.alive || !isEnemyTeam(team, other.team)) continue;
    v3set(_eye, other.position.x, other.position.y + 1.6, other.position.z);
    anglesToForward(_forward, other.yaw, other.pitch);
    if (!inCone(_eye, _forward, point, SPAWN.enemyViewConeHalfAngle, 70)) continue;
    if (!collision.isVisible(_eye, point, sightFilter)) continue;
    out.push(other.id);
  }
  return out;
}
